Create bespoke long/long hashmap and long-valued priority queue for PoolChunk

chrisvest opened the pull request:
Motivation:
The uncached access to PoolChunk can be made faster, and avoid allocating boxed Longs, if we have a primitive hash map and priority queue implementation for it.

Modification:
Add bespoke primitive implementations of a hash map and a priority queue for PoolChunk.
Remove all the long-boxing caused by the previous implementation.
The hashmap is a linear probing map with a fairly short probe that keeps the search within a couple of cache lines.
The priority queue is the same priority heap algorithm that the JDK PriorityQueue uses, but without the Long boxing, so it can instead rely on a long[] array.
This makes the internal-remove method faster, which is an important operation in PoolChunk.

Result:
I see a roughly 25% performance uplift in buffer allocations that miss cache.
(However, after #10825 fixed the caching, this boost is not so easy to measure)

normanmaurer commented:
Just allocate stuff > 32k buffers and you still will see the win

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolChunk.java:
```diff
@@ -643,4 +638,161 @@ static boolean isSubpage(long handle) {
+            }
+        }
+
+        public void remove(long key) {
```
Do we need to also shrink the internal storage at some point ?

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolChunk.java:
```diff
@@ -643,4 +638,161 @@ static boolean isSubpage(long handle) {
+            lift();
+        }
+
+        public void remove(long value) {
```
Do we also need to shrink the internal storage at some point ?

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolChunk.java:
```diff
@@ -531,8 +526,8 @@ private long collapseNext(long handle) {
-            Long nextRun = getAvailRunByOffset(runOffset + runPages);
-            if (nextRun == null) {
+            long nextRun = getAvailRunByOffset(runOffset + runPages);
+            if (nextRun == 0) {
```
```suggestion
            if (nextRun == -1) {
```
== 0 seems like a bug

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolChunk.java:
```diff
@@ -507,8 +502,8 @@ private long collapsePast(long handle) {
-            Long pastRun = getAvailRunByOffset(runOffset - 1);
-            if (pastRun == null) {
+            long pastRun = getAvailRunByOffset(runOffset - 1);
+            if (pastRun == 0) {
```
```suggestion
            if (pastRun == -1) {
```

== 0 seems incorrect

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolChunk.java:
```diff
@@ -643,4 +638,161 @@ static boolean isSubpage(long handle) {
+
+        public long poll() {
+            if (size == 0) {
+                return 0;
```
```suggestion
                return -1;
```
I think -1 is better as 0 is a valid value for `offer`

normanmaurer replied:
If we do this change we also need to adjust a few other places that check for `== 0`

franz1981 commented on buffer/src/main/java/io/netty/buffer/PoolArena.java:
```diff
@@ -190,7 +190,7 @@ private void tcacheAllocateNormal(PoolThreadCache cache, PooledByteBuf<T> buf, f
     }
 
-    // Method must be called inside synchronized(this) { ... } block
+    // Method must be called inside synchronized(this) { ... } block
```
let's put a proper `assert` with monitor check

normanmaurer replied:
Lets do this as a follow up as its not related to the PR

franz1981 reviewed (approved):
Nice one, curious about benchs, but TBH not that necessary :)

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolChunk.java:
```diff
@@ -245,18 +239,18 @@ private void insertAvailRun(int runOffset, int pages, Long handle) {
-        assert pre == null;
+    private void insertAvailRun0(int runOffset, long handle) {
+        long pre = runsAvailMap.put(runOffset, handle);
+        assert pre == 0;
```
what is the reason for returning `0` and not `-1` here ? IMHO `-1` would be what I expect for a non existing key as handles will always be positive.

chrisvest commented:
@normanmaurer I've addressed your comments. Please take a look.
@franz1981 I'll see if I can easily get this benchmarked through temporarily disabling the caching fix.

johnou reviewed (approved):
Great work

normanmaurer commented:
@chrisvest just allocate 64k buffers, this will not hit the cache by default

chrisvest commented:
@franz1981 @normanmaurer For the 64 KiB allocation case, we get a 13.2% performance increase from these changes:

```
Before:
Benchmark                                         Mode  Cnt        Score       Error  Units
PooledByteBufAllocatorBenchmark.allocateAndFree  thrpt   20  5507706.995 ± 41819.785  ops/s

After:
Benchmark                                         Mode  Cnt        Score        Error  Units
PooledByteBufAllocatorBenchmark.allocateAndFree  thrpt   20  6235283.909 ± 100380.180  ops/s
```

normanmaurer reviewed (approved):
Great work!

njhill reviewed (commented):
@chrisvest apologies for the belated review, I'd already started to look at it before it was merged.

As others have said, great work! Just had a few minor comments on the hashmap impl.. I expect it would make marginal difference if any, but was wondering if it might be possible to eliminate some branches.

chrisvest commented on buffer/src/main/java/io/netty/buffer/LongLongHashMap.java:
```diff
@@ -0,0 +1,129 @@
+    private int mask;
+    private long[] array;
+    private int maxProbe;
+    private long zeroVal;
```
Zero is used as a marker in the array, that an entry is unused. So we need to handle them specially because those keys cannot be stored in the array.

chrisvest commented on buffer/src/main/java/io/netty/buffer/LongLongHashMap.java:
```diff
@@ -0,0 +1,129 @@
+        this.emptyVal = emptyVal;
+        zeroVal = emptyVal;
+        int initialSize = 32;
+        array = new long[initialSize];
```
Branch predictors are really good these days. Meanwhile, initialising an array is not free, and would happen at resize events where we already have a spike in service time. So I think this is a better trade off.
